src/Orleans/Configuration/ServiceCollectionExtensions.cs

```diff
@@ -10,6 +10,45 @@ namespace Orleans.Configuration
     public static class ServiceCollectionExtensions
     {
         /// <summary>
+        /// Attempts to use an existing registration of <typeparamref name="TImplementation"/> to satisfy the service type <typeparamref name="TService"/>.
+        /// </summary>
+        /// <typeparam name="TService">The service type being provided.</typeparam>
+        /// <typeparam name="TImplementation">The implementation of <typeparamref name="TService"/>.</typeparam>
+        /// <param name="services">The service collection.</param>
+        internal static void AddFromExisting<TService, TImplementation>(this IServiceCollection services) where TImplementation : TService
+        {
+            var registration = services.FirstOrDefault(service => service.ServiceType == typeof(TImplementation));
+            if (registration != null)
+            {
+                var newRegistration = new ServiceDescriptor(
+                    typeof(TService),
+                    sp => sp.GetRequiredService<TImplementation>(),
+                    registration.Lifetime);
+                services.Add(newRegistration);
+            }
+        }
+
+        /// <summary>
+        /// Attempts to use an existing registration of <typeparamref name="TImplementation"/> to satisfy the service type <typeparamref name="TService"/>.
+        /// </summary>
+        /// <typeparam name="TService">The service type being provided.</typeparam>
+        /// <typeparam name="TImplementation">The implementation of <typeparamref name="TService"/>.</typeparam>
+        /// <param name="services">The service collection.</param>
+        internal static void TryAddFromExisting<TService, TImplementation>(this IServiceCollection services) where TImplementation : TService
+        {
+            var implementation = services.FirstOrDefault(service => service.ServiceType == typeof(TImplementation));
+            var providedService = services.FirstOrDefault(service => service.ServiceType == typeof(TService));
```

Comment:
AddFromExisting = "if TImplementation exists, register it as a provider of TService"
TryAddFromExisting = "if TImplementation exists and there are currently no providers of TService, register TImplementation as a provider of TService"

I could simplify the Try method to call into AddFromExisting. I'll fix the doc comments.